File: tape/sdk/typescript/src/chaos/reliability.ts
// Reliability Surface — R(k, ε, λ). Mirrors `tape.chaos.reliability`.
//
//   k  scenarios driven
//   ε  invariant-violation rate (0.0 best, 1.0 worst)
//   λ  recovery rate (1.0 best, 0.0 worst)

import type { ChaosReport } from './scenarios.ts';

export interface ReliabilitySurface {
  k: number;
  epsilon: number;
  lambda: number;
  toString(): string;
}

interface Row {
  scenarioName: string;
  passed: boolean;
  failedInvariants: string[];
  terminal: boolean;
  notes: string[];
}

export class Recorder {
  private rows: Row[] = [];

  add(report: ChaosReport, opts: { terminal?: boolean } = {}): void {
    const terminal = opts.terminal ?? true;
    const failed = (report.invariantResults ?? []).filter(ir => !ir.passed).map(ir => ir.name);
    this.rows.push({
      scenarioName: report.scenarioName,
      passed: report.passed,
      failedInvariants: failed,
      terminal,
      notes: [...(report.notes ?? [])],
    });
  }

  get surface(): ReliabilitySurface {
    const k = this.rows.length;
    if (k === 0) return makeSurface(0, 0, 1);
    const violations = this.rows.filter(r => !r.passed).length;
    const terminal = this.rows.filter(r => r.terminal).length;
    return makeSurface(k, violations / k, terminal / k);
  }

  get allRows(): readonly Row[] { return this.rows; }

  /** Render the campaign as a Markdown report — paste-into-PR friendly. */
  toMarkdown(opts: { title?: string } = {}): string {
    const s = this.surface;
    const lines: string[] = [
      `# ${opts.title ?? 'TapeChaos campaign'}`,
      '',
      `**Reliability Surface**: \`R(k=${s.k}, ε=${s.epsilon.toFixed(2)}, λ=${s.lambda.toFixed(2)})\``,
      '',
      `- ${s.k} scenarios`,
      `- ${Math.round(s.epsilon * s.k)} invariant violations`,
      `- ${Math.round(s.lambda * s.k)} runs reached terminal`,
      '',
      '| Scenario | Passed | Terminal | Failed invariants |',
      '|---|---|---|---|',
    ];
    for (const r of this.rows) {
      lines.push(
        `| \`${r.scenarioName}\` | ${r.passed ? 'OK' : 'FAIL'} | ${r.terminal ? 'yes' : 'no'} | ${r.failedInvariants.join(', ') || '—'} |`,
      );
    }
    return lines.join('\n');
  }
}

function makeSurface(k: number, epsilon: number, lambda: number): ReliabilitySurface {
  return {
    k, epsilon, lambda,
    toString() {
      return `R(k=${this.k}, ε=${this.epsilon.toFixed(2)}, λ=${this.lambda.toFixed(2)})`;
    },
  };
}

/** Quick score for a list of reports — same as `new Recorder()` + adds. */
export function score(reports: readonly ChaosReport[]): ReliabilitySurface {
  const rec = new Recorder();
  for (const r of reports) rec.add(r);
  return rec.surface;
}

File: tape/sdk/go/chaos/reliability.go
package chaos

import (
	"fmt"
	"strings"
)

// ReliabilitySurface — R(k, ε, λ). Mirrors `tape.chaos.reliability`.
//
//	k       scenarios driven
//	Epsilon invariant-violation rate
//	Lambda  recovery rate
type ReliabilitySurface struct {
	K       int
	Epsilon float64
	Lambda  float64
}

func (s ReliabilitySurface) String() string {
	return fmt.Sprintf("R(k=%d, ε=%.2f, λ=%.2f)", s.K, s.Epsilon, s.Lambda)
}

type recorderRow struct {
	ScenarioName     string
	Passed           bool
	FailedInvariants []string
	Terminal         bool
	Notes            []string
}

// Recorder — accumulator for chaos campaign results.
type Recorder struct {
	rows []recorderRow
}

// NewRecorder — fresh recorder.
func NewRecorder() *Recorder { return &Recorder{} }

// AddOpts — knobs for Recorder.Add.
type AddOpts struct {
	Terminal bool // true = the run reached a terminal state despite faults
}

// Add — record one scenario's outcome.
func (r *Recorder) Add(report ChaosReport, opts AddOpts) {
	failed := make([]string, 0)
	for _, ir := range report.InvariantResults {
		if !ir.Passed {
			failed = append(failed, ir.Name)
		}
	}
	r.rows = append(r.rows, recorderRow{
		ScenarioName:     report.ScenarioName,
		Passed:           report.Passed,
		FailedInvariants: failed,
		Terminal:         opts.Terminal,
		Notes:            append([]string(nil), report.Notes...),
	})
}

// Surface — the current R(k, ε, λ).
func (r *Recorder) Surface() ReliabilitySurface {
	k := len(r.rows)
	if k == 0 {
		return ReliabilitySurface{K: 0, Epsilon: 0, Lambda: 1}
	}
	violations := 0
	terminal := 0
	for _, row := range r.rows {
		if !row.Passed {
			violations++
		}
		if row.Terminal {
			terminal++
		}
	}
	return ReliabilitySurface{
		K:       k,
		Epsilon: float64(violations) / float64(k),
		Lambda:  float64(terminal) / float64(k),
	}
}

// ToMarkdownOpts — knobs for ToMarkdown.
type ToMarkdownOpts struct {
	Title string // default: "TapeChaos campaign"
}

// ToMarkdown — render the campaign as a Markdown table.
func (r *Recorder) ToMarkdown(opts ToMarkdownOpts) string {
	title := opts.Title
	if title == "" {
		title = "TapeChaos campaign"
	}
	s := r.Surface()
	var b strings.Builder
	fmt.Fprintf(&b, "# %s\n\n", title)
	fmt.Fprintf(&b, "**Reliability Surface**: `R(k=%d, ε=%.2f, λ=%.2f)`\n\n", s.K, s.Epsilon, s.Lambda)
	fmt.Fprintf(&b, "- %d scenarios\n", s.K)
	fmt.Fprintf(&b, "- %d invariant violations\n", int(s.Epsilon*float64(s.K)+0.5))
	fmt.Fprintf(&b, "- %d runs reached terminal\n\n", int(s.Lambda*float64(s.K)+0.5))
	fmt.Fprintln(&b, "| Scenario | Passed | Terminal | Failed invariants |")
	fmt.Fprintln(&b, "|---|---|---|---|")
	for _, row := range r.rows {
		passed := "FAIL"
		if row.Passed {
			passed = "OK"
		}
		terminal := "no"
		if row.Terminal {
			terminal = "yes"
		}
		failed := strings.Join(row.FailedInvariants, ", ")
		if failed == "" {
			failed = "—"
		}
		fmt.Fprintf(&b, "| `%s` | %s | %s | %s |\n",
			row.ScenarioName, passed, terminal, failed)
	}
	return b.String()
}

// Score — quick R(k, ε, λ) for a list of reports.
func Score(reports []ChaosReport) ReliabilitySurface {
	rec := NewRecorder()
	for _, r := range reports {
		rec.Add(r, AddOpts{Terminal: true})
	}
	return rec.Surface()
}
